_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,389 @@
+  
+  // pretend this is from an external source
+  public attack = '<img src=1 onerror="alert(\'YIKES!\')"';
+
+  constructor(private renderer: Renderer2) { }
+
+  public ngAfterViewInit(): void {
+    
+    // danger below!
+    this.el.nativeElement.innerHTML = this.attack;
+    this.renderer.setProperty(this.el.nativeElement, 'innerHTML', this.attack);
+  }
+}
+```
+
+Something like this might be tempting in a fancy custom directive, but think again! Besides, interacting directly with the DOM like this isn't the best practice in Angular, despite any security issues it might have. You should always prefer creating and using Angular templates.
+
+### Explicitly sanitize data
+
+The `DomSanitizer` class also has a method to sanitize values explicitly.
+
+Let's say you cook up a legitimate need to use `ElementRef` or `Render2` to build out the DOM in code. You can sanitize the value you add to the DOM using the method `sanitize()`. The `sanitize()` method takes two parameters, the security context for the sanitization and the value. The security context is an enumeration matching the security context listed previously.
+
+If we redo the `YikesComponent` to explicitly sanitize, the code looks like this.
+
+```ts
+@Component({
+  selector: 'app-way-better',
+  template: `
+    <div #waybetter></div>
+  `
+})
+export class WayBetterComponent implements AfterViewInit {
+
+  @ViewChild('waybetter') public el!: ElementRef<HTMLElement>;
+  
+  // pretend this is from an external source
+  public attack = '<img src=1 onerror="alert(\'YIKES!\')"';
+
+  constructor(private renderer: Renderer2, private sanitizer: DomSanitizer) { }
+
+  public ngAfterViewInit(): void {
+    
+    const cleaned = this.sanitizer.sanitize(SecurityContext.HTML, this.attack);
+    this.renderer.setProperty(this.el.nativeElement, 'innerHTML', cleaned);
+  }
+}
+```
+
+Now you'll have the image without the potentially shady code.
+
+### Consider Trusted Types
+
+One more built-in security mechanism in Angular is setting up and using a Content Security Policy (CSP). CSPs are a specific HTTP security header we covered in the [first post](first post link) to help set up foundational security mechanisms.
+
+Angular has built-in support for defining policies for a CSP called [Trusted Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types). Trusted Types are a great way to add extra XSS security guards to your Angular app but isn't supported across all the major browsers yet. If you are interested in learning more about setting up the Trusted Types CSP for SPAs, check out this great post from the Auth0 blog - [Securing SPAs with Trusted Types](https://auth0.com/blog/securing-spa-with-trusted-types/). 
+
+## Learn more about XSS, Trusted Types, and creating applications using Angular
+
+This series taught us about web security, common web attacks, and how Angular's built-in security mechanisms protect us from accidental attacks.
```

Comment:
@havi I may change this sentence depending on if I feel like I can get to a post about authorization and access control in Angular. It will depend on how much progress I make on my presentations. 😩 😰

Reply:
got it!

Reply:
I will make this a stand-alone post (if I do one), so this series will end here.